safeweb/http.go

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package safeweb
+
+import (
+	"context"
+	crand "crypto/rand"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gorilla/csrf"
+	"tailscale.com/tsnet"
+)
+
+// the default Content-Security-Policy header.
+// this policy:
+// - specifies the origin as the only valid source for all content types
+// - disallows inline javascript.
+// - disallows framing of the page
+// - disallows form submissions to other origins
+// - disallows base URIs from other origins
+// - disallows mixed content when serving over HTTPS
+// - disallows embedding of resources from other origins
+const defaultCSP string = `
+	default-src 'self';
+	script-src 'self';
+	frame-ancestors 'none';
+	form-action 'self';
+	base-uri 'self';
+	block-all-mixed-content;
+	object-src 'none'
+`
+
+type Config struct {
+	// Hostname is the name of the tsnet service that will be created to host
+	// the application.
+	Hostname string
+
+	// Listener is a net.Listener that the server will host the application on.
+	// Providing a listening will override the use of tsnet and the Hostname field.
+	Listener net.Listener
+
+	// BrowserMux is the HTTP handler for any routes in your application
+	// that should be served to browsers. These requests will be subject to CSRF
+	// protection and will have browser-specific headers set in response.
+	BrowserMux *http.ServeMux
+
+	// APIMux is the HTTP handler for any routes in your application that
+	// should be served to non-browser clients.
+	APIMux *http.ServeMux
+
+	// PermittedCrossOriginHosts specifies the Access-Control-Allow-Origin header sent in response to pre-flight OPTIONS requests.
```

Comment:
We should document what we accept here. E.g., do we allow anything the Access-Control-Allow-Origin header does, including the wildcard?

If it's the same syntax, maybe we should use that name, `AllowOrigin` so that the search term will line up. (No objection to "permitted" as such, but it's another synonym to keep track of)

Reply:
Please also say what an empty value means, the defaults, etc.